Monotonic sequences in posets

malarbol opened the pull request:
This PR introduces a few definitions and results to work with sequences is preorders, posets or strictly preordered sets.
As an application, it introduces the concept of `subsequences` of a sequence `u`: the sequences `u ∘ f` for some strictly increasing map `f : ℕ → ℕ`.

malarbol commented:
There's still a lot to do but if anyone sees anything they dislike, we can start talking about it.
For example, should I use `nondecreasing / increasing` instead of `increasing / strictly increasing` ?

fredrik-bakke commented on src/order-theory/sequences-posets.lagda.md:
```diff
@@ -0,0 +1,94 @@
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  poset-sequence-Poset : Poset l1 l2
```
```suggestion
  sequence-Poset : Poset l1 l2
```

fredrik-bakke commented on src/order-theory/sequences-posets.lagda.md:
```diff
@@ -0,0 +1,94 @@
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  sequence-Poset : UU l1
```
```suggestion
  type-sequence-Poset : UU l1
```

fredrik-bakke commented on src/order-theory/increasing-sequences-posets.lagda.md:
```diff
@@ -0,0 +1,100 @@
+  {l1 l2 : Level} (P : Poset l1 l2) {u : sequence-Poset P}
+  where
+
+  is-increasing-sequence-all-increasing-value-sequence-Poset :
+    ((n : ℕ) → is-increasing-value-sequence-Poset P u n) →
+    is-increasing-sequence-Poset P u
+  is-increasing-sequence-all-increasing-value-sequence-Poset H p =
+    based-ind-ℕ
+      ( p)
+      ( λ q → leq-Poset P (u p) (u q))
+      ( refl-leq-Poset P (u p))
+      ( λ q I →
+        transitive-leq-Poset P (u p) (u q) (u (succ-ℕ q)) (H q))
```
This looks like an instantiation of `preserves-order-ind-ℕ-Poset`

```suggestion
  is-increasing-sequence-all-increasing-value-sequence-Poset :
    ((n : ℕ) → is-increasing-value-sequence-Poset P u n) →
    is-increasing-sequence-Poset P u
  is-increasing-sequence-all-increasing-value-sequence-Poset H =
    preserves-order-ind-ℕ-Poset P u H
```

fredrik-bakke replied:
The definition is given in `elementary-number-theory.decidable-total-order-natural-numbers`

fredrik-bakke commented on src/order-theory/increasing-sequences-posets.lagda.md:
```diff
@@ -0,0 +1,100 @@
+  {l1 l2 : Level} (P : Poset l1 l2) (u : sequence-Poset P) (n : ℕ)
+  where
+
+  is-increasing-value-prop-sequence-Poset : Prop l2
+  is-increasing-value-prop-sequence-Poset =
+    leq-prop-Poset P (u n) (u (succ-ℕ n))
```
I find the terminology "increasing value" awkward. A value can't be increasing, but a sequence may be. Maybe we can do away with this definition?

malarbol commented on src/order-theory/sequences-posets.lagda.md:
````diff
@@ -0,0 +1,94 @@
+  leq-value-sequence-Poset = type-Prop ∘ leq-value-prop-sequence-Poset
+
+  leq-prop-sequence-Poset : Prop l2
+  leq-prop-sequence-Poset = Π-Prop ℕ leq-value-prop-sequence-Poset
+
+  leq-sequence-Poset : UU l2
+  leq-sequence-Poset = type-Prop leq-prop-sequence-Poset
+
+  is-prop-leq-sequence-Poset : is-prop leq-sequence-Poset
+  is-prop-leq-sequence-Poset = is-prop-type-Prop leq-prop-sequence-Poset
+```
+
+## Properties
+
+### The type of sequences in a poset is a poset ordered by pointwise comparison
+
+```agda
+module _
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  sequence-Poset : Poset l1 l2
+  pr1 (pr1 sequence-Poset) = type-sequence-Poset P
+  pr1 (pr2 (pr1 sequence-Poset)) = leq-prop-sequence-Poset P
+  pr1 (pr2 (pr2 (pr1 sequence-Poset))) u n = refl-leq-Poset P (u n)
+  pr2 (pr2 (pr2 (pr1 sequence-Poset))) u v w J I n =
+    transitive-leq-Poset P (u n) (v n) (w n) (J n) (I n)
+  pr2 sequence-Poset u v I J =
+    eq-htpy (λ n → antisymmetric-leq-Poset P (u n) (v n) (I n) (J n))
+
+  refl-leq-sequence-Poset : is-reflexive (leq-sequence-Poset P)
+  refl-leq-sequence-Poset = refl-leq-Poset sequence-Poset
+
+  transitive-leq-sequence-Poset : is-transitive (leq-sequence-Poset P)
+  transitive-leq-sequence-Poset = transitive-leq-Poset sequence-Poset
+
+  antisymmetric-leq-sequence-Poset : is-antisymmetric (leq-sequence-Poset P)
+  antisymmetric-leq-sequence-Poset =
+    antisymmetric-leq-Poset sequence-Poset
+```
````
Is this covered anywhere in the order-theory module? e.g. products of posets or anything?

fredrik-bakke replied:
We have `order-theory.dependent-products-large-posets`, so what you should do is copy its contents to a new file `order-theory.dependent-products-posets` and change it to apply to (small) posets. Note that if you want to eventually apply this theory of sequences in posets to real numbers, you should use the large variant instead.

malarbol replied:
ok. Since I'm mainly focused on sequences of natural numbers (to define subsequences) and rational numbers (to talk about monotonicity of arithmetic / geometric sequences in Q+) can we leave this for another PR? Then, when we work on monotonic sequences of real numbers we'll do sequences in Large Posets an maybe refactor this?

malarbol replied:
Are the "strictly preordered sets" of natural numbers and rational numbers defined anywhere? I'll need them to define strictly increasing sequences in term of "strict order preserving map".

fredrik-bakke replied:
The strict ordering is defined, but probably not that they are strictly ordered sets since the addition of strictly ordered sets is very new.

malarbol commented on src/elementary-number-theory/strict-inequality-natural-numbers.lagda.md:
```diff
@@ -148,6 +151,15 @@ transitive-le-ℕ (succ-ℕ n) (succ-ℕ m) (succ-ℕ l) p q =
+ℕ-Strict-Preordered-Set : Strictly-Preordered-Set lzero lzero
+pr1 ℕ-Strict-Preordered-Set = ℕ-Set
+pr2 ℕ-Strict-Preordered-Set =
+  le-ℕ-Prop , anti-reflexive-le-ℕ , λ n m l I J → transitive-le-ℕ n m l J I
```
If I understood correctly `le-ℕ-Prop` should be renamed `le-prop-ℕ`. Do you want me to fix this now?

malarbol commented on src/elementary-number-theory/strict-inequality-natural-numbers.lagda.md:
```diff
@@ -148,6 +151,15 @@ transitive-le-ℕ (succ-ℕ n) (succ-ℕ m) (succ-ℕ l) p q =
+ℕ-Strict-Preordered-Set : Strictly-Preordered-Set lzero lzero
+pr1 ℕ-Strict-Preordered-Set = ℕ-Set
+pr2 ℕ-Strict-Preordered-Set =
+  le-ℕ-Prop , anti-reflexive-le-ℕ , λ n m l I J → transitive-le-ℕ n m l J I
```
There's also a bit of inconsistency in the names, between `anti-reflexive` here and `irreflexive` in [order-theory.strictly-preordered-sets](https://unimath.github.io/agda-unimath/order-theory.strictly-preordered-sets.html). How do you want to handle this? (same for the order of arguments for transitivity).

malarbol replied:
> Presumably, Egbert intentionally introduced "anti-reflexive", so you can keep that if you want. There's nothing wrong with `irreflexive` though.

Sure, I meant that `anti-reflexive-le-ℕ` could be renamed `irreflexive-le-ℕ` but, as for `le-prop-ℕ`, I'll leave it this way for now and you can decide later want you want to do with this.

malarbol commented on src/order-theory/sequences-posets.lagda.md:
```diff
@@ -0,0 +1,94 @@
+  {l1 l2 : Level} (P : Poset l1 l2)
+  where
+
+  sequence-Poset : Poset l1 l2
```
see https://github.com/UniMath/agda-unimath/pull/1388/commits/511889ab909cacbf6c0f27756bf14285c72a5c22
I guess we could do something like `sequence-Preorder` too but I'd rather keep this for a further re-factor if everyone is ok with it.

malarbol replied:
> see [511889a](https://github.com/UniMath/agda-unimath/commit/511889ab909cacbf6c0f27756bf14285c72a5c22) I guess we could do something like `sequence-Preorder` too but I'd rather keep this for a further re-factor if everyone is ok with it.

I think I'll go for it. Maybe I'll even try sequences in large preorders.

malarbol commented on src/order-theory/strictly-increasing-sequences-strictly-preordered-sets.lagda.md:
````diff
@@ -0,0 +1,114 @@
+```agda
+module _
+  {l1 l2 : Level} (A : Strictly-Preordered-Set l1 l2)
+  where
+
+  is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set :
+    (u : type-sequence-Strictly-Preordered-Set A) →
+    ((n : ℕ) → le-Strictly-Preordered-Set A (u n) (u (succ-ℕ n))) →
+    is-strictly-increasing-sequence-Strictly-Preordered-Set A u
+  is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set
+    u H zero-ℕ (succ-ℕ zero-ℕ) I = H zero-ℕ
+  is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set
+    u H zero-ℕ (succ-ℕ (succ-ℕ n)) I =
+      is-transitive-le-Strictly-Preordered-Set A
+        ( u zero-ℕ)
+        ( u (succ-ℕ n))
+        ( u (succ-ℕ (succ-ℕ n)))
+        ( H (succ-ℕ n))
+        ( is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set
+          ( u)
+          ( H)
+          ( zero-ℕ)
+          ( succ-ℕ n)
+          ( I))
+  is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set
+    u H (succ-ℕ m) (succ-ℕ n) I =
+      is-strictly-increasing-is-everywhere-strictly-increasing-sequence-Strictly-Preordered-Set
+        ( u ∘ succ-ℕ)
+        ( H ∘ succ-ℕ)
+        ( m)
+        ( n)
+        ( I)
+
+  is-everywhere-strictly-increasing-is-strictly-increasing-sequence-Strictly-Preordered-Set :
+    (u : type-sequence-Strictly-Preordered-Set A) →
+    is-strictly-increasing-sequence-Strictly-Preordered-Set A u →
+    ((n : ℕ) → le-Strictly-Preordered-Set A (u n) (u (succ-ℕ n)))
+  is-everywhere-strictly-increasing-is-strictly-increasing-sequence-Strictly-Preordered-Set
+    u H n = H n (succ-ℕ n) (succ-le-ℕ n)
+```
````
This is meant to address https://github.com/UniMath/agda-unimath/pull/1378#discussion_r2012480612
The name is a bit long but I'm open for suggestions.

lowasser commented:
It looks like many of my comments from #1371 apply to this subset.  I'm going to approve this, but please apply the changes asked for there to this PR.

malarbol commented:
> It looks like many of my comments from #1371 apply to this subset. I'm going to approve this, but please apply the changes asked for there to this PR.

I think I addressed all your comments from #1371 specific to this sub-PR. You can merge it or we can do it all in #1371 if you prefer.

lowasser commented:
I don't have merging permissions at this point, just review.

malarbol commented:
> I don't have merging permissions at this point, just review.

ok. we'll wait for @fredrik-bakke then. Thanks a lot for your help! See you on #1371, #1378, #1398 and, hopefully, more.

malarbol commented:
Is anyone interested in merging this separately from #1371? Otherwise I'll just close it and we can go directly to #1371.

fredrik-bakke commented on src/foundation/strictly-right-unital-concatenation-identifications.lagda.md:
````diff
@@ -64,7 +64,7 @@ module _
 ```
 
-### Translating between the stictly left and stricly right unital versions of concatenation
+### Translating between the strictly left and strictly right unital versions of concatenation
````
Wow, I managed to misspell it in two different ways in the same header 😅 thanks for fixing it!

malarbol replied:
For some reason I misspell this word a lot too so this file just randomly popped up doing a `grep`. I was indeed surprised to see it pop up twice 🤣.

fredrik-bakke replied:
You're not alone apparantly 🙂

fredrik-bakke commented on src/order-theory/infinite-limit-sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,136 @@
+      (modulus-modulus-limit-∞-sequence-Preorder m)
+  is-modulus-modulus-limit-∞-sequence-Preorder = pr2
+
+  is-limit-∞-prop-sequence-Preorder : Prop (l1 ⊔ l2)
```
Have you considered the perhaps more idiomatic `tends-to`? Then this could simply be read off as `tends-to-infinity`.

malarbol replied:
ok, I guess it will go better with https://github.com/UniMath/agda-unimath/pull/1388/#discussion_r2040678207 too. I'll fix that, and I guess zero-approximations of #1371 will `tend-to-zero` instead.

fredrik-bakke commented on src/order-theory/infinite-limit-sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,136 @@
+    modulus-limit-∞-sequence-Preorder → type-Preorder P → ℕ
+  modulus-modulus-limit-∞-sequence-Preorder = pr1
+
+  is-modulus-modulus-limit-∞-sequence-Preorder :
```
We prefer the textual `infinity` to the unicode symbol `∞` inside of entry names.

https://unimath.github.io/agda-unimath/CODINGSTYLE.html#naming-conventions

fredrik-bakke commented on src/order-theory/infinite-limit-sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,136 @@
+  {l1 l2 : Level} (P : Preorder l1 l2)
+  where
+
+  limit-∞-sequence-Preorder : UU (l1 ⊔ l2)
```
Potentially `sequence-tending-to-infinity-Preorder`

malarbol replied:
I went with `type-tends-to-infinity-sequence-Preorder`. Is it ok?

fredrik-bakke replied:
How is that better?

malarbol replied:
Does the equation 
```
type-tends-to-infinity-sequence-Preorder =
    type-subtype (subtype-tends-to-infinity-sequence-Preorder P)
```
convince you? with `-subtype subtype-` cancelling out? 
Furthermore, let's say we rename `sequence-Preorder` to `preorder-sequence-Preorder` as I suggested [there](https://github.com/UniMath/agda-unimath/pull/1388#discussion_r2040694185) and we have something like https://github.com/UniMath/agda-unimath/pull/1371/commits/2cba40a80a2f96913ed1c0ad714a4269995b3524 (I could include here it if you'd like, it actually fits beer here than in #1371) we could write
```
preorder-tends-to-infinity-sequence-Preorder : Preorder _ _
preorder-tends-to-infinity-sequence-Preorder = 
  preorder-subtype-Preorder (preorder-sequence-Preorder P) (subptype-tends-to-infinity-sequence-Preorder P)
```
so
```
type-Preorder (preorder-tends-to-infinity-sequence-Preorder P) == type-tends-to-infinity-sequence-Preorder P
```
etc.

malarbol replied:
anyways, I just realizes https://github.com/UniMath/agda-unimath/commit/2cba40a80a2f96913ed1c0ad714a4269995b3524 is mostly `Subposets` and I got all the names wrong. I'll have another look.

fredrik-bakke commented on src/order-theory/increasing-sequences-posets.lagda.md:
```diff
@@ -0,0 +1,74 @@
+
+## Properties
+
+### A sequence `u` in a poset is increasing if and only if `uₙ ≤ uₙ₊₁` for all n : ℕ
```
```suggestion
### A sequence `u` in a poset is increasing if and only if `uₙ ≤ uₙ₊₁` for all `n : ℕ`
```

fredrik-bakke commented on src/foundation/subsequences.lagda.md:
````diff
@@ -0,0 +1,156 @@
+### Subsequences of a sequence
+
+```agda
+module _
+  {l : Level} {A : UU l} (u : sequence A)
+  where
+
+  subsequence : UU lzero
+  subsequence =
+    hom-Strictly-Preordered-Set
+      strictly-preordered-set-ℕ
+      strictly-preordered-set-ℕ
````
It seems like it could be useful to show that this is equivalent to unbounded (decidable?) subsets of the natural numbers. Just an idea

malarbol replied:
I'll give it some thoughts, but I already have a lot of plans 😅 . Is there anything already on (un)bounded subsets of posets?

fredrik-bakke replied:
don't think so

fredrik-bakke commented on src/foundation/subsequences.lagda.md:
```diff
@@ -0,0 +1,156 @@
+
+  limit-∞-extract-subsequence : limit-∞-sequence-Preorder ℕ-Preorder
+  limit-∞-extract-subsequence =
+    extract-subsequence u v , is-limit-∞-extract-subsequence
```
We parenthesize top-level pairs to be consistent with the mathematical notation. (this convention is currently undocumented, sorry)

```suggestion
    ( extract-subsequence u v , is-limit-∞-extract-subsequence)
```

fredrik-bakke reviewed (approved):
Very nice additions, thank you for these contributions!

fredrik-bakke commented:
Is it problematic to merge this PR without #1371? It is helpful to me to portion the review job out a bit (2700 lines is a lot for me)

malarbol commented on src/order-theory/sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,89 @@
+  transitive-leq-sequence-Preorder u v w J I n =
+    transitive-leq-Preorder P (u n) (v n) (w n) (J n) (I n)
+
+  sequence-Preorder : Preorder l1 l2
```
Should this be `preorder-sequence-Preorder`, so we'd have 
```
type-Preorder (preorder-sequence-Preorder P) = type-sequence-Preorder
```
with like `Preorder preorder` cancelling out?

fredrik-bakke replied:
No, this should be `sequence-Preorder`

fredrik-bakke replied:
The underlying type can be `type-sequence-Preorder`

fredrik-bakke commented on src/order-theory/infinite-limit-sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,143 @@
+  {l1 l2 : Level} (P : Preorder l1 l2)
+  where
+
+  type-tends-to-infinity-sequence-Preorder : UU (l1 ⊔ l2)
+  type-tends-to-infinity-sequence-Preorder =
+    type-subtype (subtype-tends-to-infinity-sequence-Preorder P)
```
Wait, this name doesn't make sense at all. It is the underlying type of the preorder of sequences that tend to infinity, so the proper name is probably `type-sequence-tending-to-infinity-Preorder`

fredrik-bakke replied:
It is not the underlying type of the preorder of witnesses that a given sequence in a preorder tends to infinity, as the name suggests.

malarbol replied:
What about `type-tending-to-infinity-sequence-Preorder`? it seems more coherent with names like `type-increasing-sequence-Preorder` etc., considering `tending-to-infinity` as a whole adjective-like term?

fredrik-bakke replied:
I have to say the phrasing a "tending-to-infinity-sequence" looks very odd. Have you ever heard anyone say that or seen it used in writing?

malarbol replied:
I agree it's very Yoda. But, from a library POV, it really feels that this should live in the namespace `XXX-sequence-Preorder`. I wish there was an adjective for "tending to infinity" but it's the best thing I have.

malarbol replied:
Maybe it's just I defect of the English language? For some reason, you can say "This sequence is increasing, it is an increasing sequence", but when I say, "This sequence is tending to infinity, it is a tending to infinity sequence", I sound lunatic 🤣

fredrik-bakke replied:
"Diverging" is different because a sequence can be bounded and still diverge. Infinitely diverging is not correct nomenclature. Upward diverging is also bad terminology, I believe, because the dual concept is convergence, not downward divergence.

fredrik-bakke replied:
You could maybe call it unbounded divergence, which consistently suggests "modulus of unbounded divergence", and then, the sequence would be called an "unbounded sequence"

fredrik-bakke replied:
Does that conflict with a different related concept?

malarbol replied:
> "Diverging" is different because a sequence can be bounded and still diverge. 

Sure, that's why it has to be `XXX-diverging` where `XXX` states its an diverging to infinity.

> Infinitely diverging is not correct nomenclature. Upward diverging is also bad terminology

ok.But it still seems that the phrasing "diverges to infinity" is more used that "converges to infinity", and avoid the conflict with convergence in metric spaces, where "converges to infinity" will have a totally different definition.

fredrik-bakke replied:
Note that this seems to me to be different to "tending to infinity", which should be akin to convergence to the point at infinity in the 1-point compactification.

fredrik-bakke replied:
E.g., the sequence `1,0,2,0,3,0,4,0,5,0,6,0,7,0,8...` does not tend to infinity, since we can establish a positive distance between infinity and "all elements above n". However, the sequence `1,2,3,4,5,6,7,8,...` does tend to infinity.

fredrik-bakke replied:
I don't have more time to discuss this now. Please figure out the terminology, then I can give it a new review in about 10 days.

malarbol replied:
> You could maybe call it unbounded divergence, which consistently suggests "modulus of unbounded divergence", and then, the sequence would be called an "unbounded sequence"

unbounded sequences is yet another concept. a sequence can be unbounded and not diverge to infinity (e.g. ` n -> n*(-1)^n`)

malarbol replied:
> I don't have more time to discuss this now. Please figure out the terminology, then I can give it a new review in about 10 days.

ok. thanks a lot for your time.

malarbol replied:
I removed all the `tending-to-infinity` stuff for now. I'll want it for things like "subsequences preserve limits" or "arithmetic sequences in Q+ diverge to infinity" but I'll give it more thoughts and put It in #1371 or another intermediate PR if needed. If the rest is ok for you, we can merge it.

fredrik-bakke commented on src/order-theory/infinite-limit-sequences-preorders.lagda.md:
`sequences-tending-to-infinity-preorders.lagda.md`

fredrik-bakke commented:
> I added a few modules on bounded sequences. I still haven't find a good name for sequences tending to infinity so I'd rather leave this for later. Otherwise, I think it's good to go.

Were these files suggested in the previous review?
If not, please pull them out to a second PR so this one can be merged.

malarbol commented:
> Were these files suggested in the previous review? If not, please pull them out to a second PR so this one can be merged.

Not really. At some point I wanted to use them to talk about unbounded sequences / sequences tending to infinity but I didn't went through. They're out now.

malarbol commented on src/order-theory/sequences-preorders.lagda.md:
```diff
@@ -0,0 +1,89 @@
+  {l1 l2 : Level} (P : Preorder l1 l2)
+  where
+
+  type-sequence-Preorder : UU l1
+  type-sequence-Preorder = sequence (type-Preorder P)
```
Now that I see it, `sequence-type-Preorder` seems much better, and quite obvious. Let me now when/how you want to fix this.

fredrik-bakke replied:
If you want you can submit a pull request with just that naming change

malarbol replied:
see #1407
